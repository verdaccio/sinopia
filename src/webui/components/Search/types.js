/**
 * @prettier
 * @flow
 */

export interface IProps {
  history?: any;
}

export interface IState {
  search: string;
  suggestions: any[];
  loading: boolean;
  loaded: boolean;
  error: boolean;
}

export type cancelAllSearchRequests = () => void;
export type handlePackagesClearRequested = () => void;
export type handleSearch = (event: SyntheticKeyboardEvent<HTMLInputElement>, { newValue: string, method: string }) => void;
export type handleClickSearch = (event: SyntheticKeyboardEvent<HTMLInputElement>, { suggestionValue: Array<Object>, method: string }) => void;
export type handleFetchPackages = ({ value: string }) => Promise<void>;
export type onBlur = (event: SyntheticKeyboardEvent<HTMLInputElement>) => void;
