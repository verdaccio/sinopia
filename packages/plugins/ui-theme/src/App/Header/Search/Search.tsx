import debounce from 'lodash/debounce';
import React, { FormEvent, useCallback, useState } from 'react';
import { SuggestionSelectedEventData } from 'react-autosuggest';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { RouteComponentProps, withRouter } from 'react-router';
import AutoComplete from 'verdaccio-ui/components/AutoComplete';

import { Dispatch, RootState } from '../../../store/store';
import SearchAdornment from './SearchAdornment';

const CONSTANTS = {
  API_DELAY: 300,
  ABORT_ERROR: 'AbortError',
};

const Search: React.FC<RouteComponentProps> = ({ history }) => {
  const { t } = useTranslation();
  const [loaded, setLoaded] = useState(false);
  const [search, setSearch] = useState('');
  // const mountedRef = useRef(true);
  const { isError, suggestions } = useSelector((state: RootState) => state.search);
  const isLoading = useSelector((state: RootState) => state?.loading?.models.search);
  const dispatch = useDispatch<Dispatch>();
  /**
   * Cancel all the requests which are in pending state.
   */
  const cancelAllSearchRequests = useCallback(() => {
    dispatch.search.clearRequestQueue();
  }, [dispatch]);

  /**
   * As user focuses out from input, we cancel all the request from requestList
   * and set the API state parameters to default boolean values.
   */
  const handleOnBlur = useCallback(
    (event: FormEvent<HTMLInputElement>) => {
      // stops event bubbling
      event.stopPropagation();
      cancelAllSearchRequests();
    },
    [cancelAllSearchRequests]
  );

  /**
   * onChange method for the input element.
   */
  const handleSearch = useCallback(
    (event: FormEvent<HTMLInputElement>, { newValue, method }) => {
      // stops event bubbling
      event.stopPropagation();
      if (method === 'type') {
        const value = newValue.trim();
        setSearch(value);
        setLoaded(false);
        /**
         * A use case where User keeps adding and removing value in input field,
         * so we cancel all the existing requests when input is empty.
         */
        if (value?.length === 0) {
          cancelAllSearchRequests();
        }
      }
    },
    [cancelAllSearchRequests]
  );

  /**
   * Cancel all the request from list and make request list empty.
   */
  const handlePackagesClearRequested = useCallback(() => {
    dispatch.search.saveSearch({ suggestions: [] });
  }, [dispatch]);

  /**
   * When an user select any package by clicking or pressing return key.
   */
  const handleClickSearch = useCallback(
    (
      event: FormEvent<HTMLInputElement>,
      { suggestionValue, method }: SuggestionSelectedEventData<unknown>
    ): void | undefined => {
      // stops event bubbling
      event.stopPropagation();
      switch (method) {
        case 'click':
        case 'enter':
          setSearch('');
          history.push(`/-/web/detail/${suggestionValue}`);
          break;
      }
    },
    [history]
  );

  /**
   * Fetch packages from API.
   * For AbortController see: https://developer.mozilla.org/en-US/docs/Web/API/AbortController
   */
  const handleFetchPackages = useCallback(
    ({ value }: { value: string }) => {
      dispatch.search.getSuggestions({ value });
    },
    [dispatch]
  );

  return (
    <AutoComplete
      onBlur={handleOnBlur}
      onChange={handleSearch}
      onCleanSuggestions={handlePackagesClearRequested}
      onClick={handleClickSearch}
      onSuggestionsFetch={debounce(handleFetchPackages, CONSTANTS.API_DELAY)}
      placeholder={t('search.packages')}
      startAdornment={<SearchAdornment />}
      suggestions={suggestions}
      suggestionsError={isError}
      suggestionsLoaded={loaded}
      suggestionsLoading={isLoading}
      value={search}
    />
  );
};

export default withRouter(Search);
