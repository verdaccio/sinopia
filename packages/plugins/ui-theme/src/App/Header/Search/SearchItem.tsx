/* eslint-disable verdaccio/jsx-spread */
import styled from '@emotion/styled';
import Cached from '@mui/icons-material/Cached';
import HttpsIcon from '@mui/icons-material/Https';
import SyncAlt from '@mui/icons-material/SyncAlt';
import { Theme } from '@mui/material';
import Chip from '@mui/material/Chip';
import Stack from '@mui/material/Stack';
import React from 'react';
import { useTranslation } from 'react-i18next';

type SearchItemProps = {
  name: string;
  version?: string;
  description?: string;
  isPrivate?: boolean;
  isCached?: boolean;
  isRemote?: boolean;
};

const Wrapper = styled.div({
  display: 'flex',
  alignItems: 'center',
  width: '100%',
});

export const Description = styled('div')<{ theme?: Theme }>(({ theme }) => ({
  display: 'none',
  color: theme?.palette?.greyLight2,
  lineHeight: '1.5rem',
  [`@media (min-width: ${theme?.breakPoints.large}px)`]: {
    display: 'block',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    width: '200px',
    alignItems: 'center',
    overflow: 'hidden',
    paddingLeft: theme.spacing(),
    fontSize: theme?.fontSize.ssm,
  },
}));

const NameGroup = styled.span({
  display: 'flex',
  flex: '1',
});

const Name = styled('span')<{ theme?: Theme }>(({ theme }) => ({
  fontWeight: '700',
  fontSize: theme?.fontSize.sm,
}));

const Version = styled('span')<{ theme?: Theme }>(({ theme }) => ({
  fontSize: theme?.fontSize.ssm,
}));

const SearchItem: React.FC<SearchItemProps> = ({
  name,
  description,
  isPrivate = false,
  isRemote = false,
  isCached = false,
  version,
  ...props
}) => {
  const { t } = useTranslation();
  const handleDelete = () => {
    // no action assigned by default
  };
  return (
    <li {...props} style={{ flexDirection: 'column' }}>
      <Wrapper>
        <NameGroup>
          <Name>{name}</Name>
          {description && <Description>{description}</Description>}
        </NameGroup>
        {version && <Version>{version}</Version>}
      </Wrapper>
      <Wrapper>
        <Stack direction="row" spacing={1}>
          {isPrivate && (
            <Chip
              color="primary"
              deleteIcon={<HttpsIcon />}
              label={t('search.isPrivate')}
              onDelete={handleDelete}
              size="small"
            />
          )}
          {isRemote && !isPrivate && (
            <Chip
              deleteIcon={<SyncAlt />}
              label={t('search.isRemote')}
              onDelete={handleDelete}
              size="small"
              variant="outlined"
            />
          )}
          {isCached && (
            <Chip
              deleteIcon={<Cached />}
              label={t('search.isCached')}
              onDelete={handleDelete}
              size="small"
              variant="outlined"
            />
          )}
        </Stack>
      </Wrapper>
    </li>
  );
};

export default SearchItem;
