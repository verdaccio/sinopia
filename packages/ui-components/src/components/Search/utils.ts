function removeHtmlTags(input: string): string {
  let previous;
  do {
    previous = input;
    input = input.replace(/<[^>]*>?/gm, '');
  } while (input !== previous);
  return input;
}

export function cleanDescription(description: string): string {
  let output = description;
  // remove html tags from description (e.g. <h1...>)
  output = removeHtmlTags(output);
  // remove markdown links from description (e.g. [link](url))
  output = output.replace(/\(.*?\)/gm, '').replace(/(\[!?|\])/gm, '');
  return output;
}
