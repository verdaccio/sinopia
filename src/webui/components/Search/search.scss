@import '../../styles/mixins';

.searchBox {
  @include searchBox;
}
