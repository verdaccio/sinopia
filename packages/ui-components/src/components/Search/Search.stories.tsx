/* eslint-disable verdaccio/jsx-spread */
import Box from '@mui/material/Box';
import React from 'react';
import { MemoryRouter } from 'react-router';

import Search from './Search';

export default {
  title: 'Components/Header/Search',
};

export const SearchByQuery = {
  render: (args) => (
    <MemoryRouter initialEntries={[`/-/web/detail/storybook`]}>
      <Box
        sx={{
          height: 100,
          padding: 10,
          backgroundColor: 'primary.dark',
          '&:hover': {
            backgroundColor: 'primary.main',
            opacity: [0.9, 0.8, 0.7],
          },
        }}
      >
        <Search {...args} />
      </Box>
    </MemoryRouter>
  ),
};
