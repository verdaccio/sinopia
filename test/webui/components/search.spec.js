/**
 * Search component
 */

import React from 'react';
import { shallow } from 'enzyme';
import Search from '../../../src/webui/src/components/Search/index';

console.error = jest.fn();

describe('<Search /> component', () => {
  it('should give error for the required fields', () => {
    const wrapper = shallow(<Search />);
    expect(console.error).toBeCalled();
    expect(wrapper
        .find('input')
        .prop('placeholder')).toEqual('Type to search...');
  });

  it('should have <input /> element with correct properties', () => {
    const props = {
      handleSearchInput: () => {},
      placeHolder: 'Test placeholder'
    };
    const wrapper = shallow(<Search {...props} />);
    expect(wrapper.find('input')).toHaveLength(1);
    expect(wrapper.find('input').prop('placeholder')).toEqual(
      'Test placeholder'
    );
    expect(typeof wrapper
        .find('input')
        .prop('onChange')).toBe('function');
  });

  it('should call the handleSearchInput function', () => {
    const props = {
      handleSearchInput: jest.fn(),
      placeHolder: 'Test placeholder'
    };
    const wrapper = shallow(<Search {...props} />);
    wrapper.find('input').simulate('change');
    expect(props.handleSearchInput).toBeCalled();
  });
});