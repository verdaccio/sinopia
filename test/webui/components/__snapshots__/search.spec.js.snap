// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> component should match the snapshot 1`] = `"<input type=\\"text\\" placeholder=\\"Type to search...\\" class=\\"searchBox\\"/>"`;
